contracts/locked-astroport-vault/src/state.rs

```diff
@@ -44,6 +44,102 @@ pub fn claims() -> Claims<'static> {
+            return Err(StdError::generic_err("Fee rate can't be higher than 100%"));
+        }
+        // If fee rate is not zero, then there must be some fee recipients and their
+        // weights must sum to 100%
+        if !self.fee_rate.is_zero()
+            && self.fee_recipients.iter().map(|(_, p)| p).sum::<Decimal>() != Decimal::one()
+        {
+            return Err(StdError::generic_err(
+                "Sum of fee recipient percentages must be 100%",
+            ));
+        }
+        Ok(FeeConfig {
+            fee_rate: self.fee_rate,
+            fee_recipients: self
+                .fee_recipients
+                .iter()
+                .map(|(addr, percentage)| Ok((deps.api.addr_validate(addr)?, *percentage)))
+                .collect::<StdResult<Vec<_>>>()?,
+        })
+    }
+}
+
+impl FeeConfig<Addr> {
+    /// Creates messages to transfer an `AssetList` of assets to the fee
+    /// recipients.
+    pub fn transfer_assets_msgs(&self, assets: &AssetList, env: &Env) -> StdResult<Vec<CosmosMsg>> {
+        Ok(self
+            .fee_recipients
+            .iter()
+            // Filter out the contract address because it's unnecessary to send fees to ourselves
+            .filter(|(addr, _)| addr != env.contract.address)
+            .map(|(addr, percentage)| {
+                let assets: AssetList = assets
+                    .iter()
+                    .map(|asset| Asset::new(asset.info.clone(), asset.amount * *percentage))
+                    .collect::<Vec<_>>()
+                    .into();
+                assets.transfer_msgs(addr)
+            })
+            .collect::<StdResult<Vec<_>>>()?
+            .into_iter()
+            .flatten()
+            .collect())
+    }
+
+    /// Calculates the fee from the input assets and returns messages to send
+    /// them to the fee recipients.
+    pub fn fee_msgs_from_assets(&self, assets: &AssetList, env: &Env) -> StdResult<Vec<CosmosMsg>> {
+        // Take fee from input assets
+        let assets: AssetList = assets
+            .iter()
+            .map(|asset| Asset::new(asset.info.clone(), asset.amount * self.fee_rate))
+            .collect::<Vec<_>>()
+            .into();
+        // Send fee to fee recipients
+        self.transfer_assets_msgs(&assets, env)
+    }
+
+    /// Calculates the fee from the input asset and returns messages to send it
+    /// to the fee recipients.
+    pub fn fee_msgs_from_asset(&self, asset: Asset, env: &Env) -> StdResult<Vec<CosmosMsg>> {
+        self.fee_msgs_from_assets(&AssetList::from(vec![asset]), env)
+    }
+}
+
+impl From<FeeConfig<Addr>> for FeeConfig<String> {
+    fn from(value: FeeConfig<Addr>) -> Self {
+        Self {
+            fee_rate: value.fee_rate,
+            fee_recipients: value
+                .fee_recipients
+                .into_iter()
+                .map(|(addr, percentage)| (addr.to_string(), percentage))
+                .collect(),
+        }
+    }
+}
```

Comment:
add some unit tests for `FeeConfig` functionality

Reply:
done